### ISSUE TITLE
Question implementation of `warp to vessel@host`

### ISSUE THREAD

**Issue by microlith57** [XXIIVV/Parade#2]
I have been implementing client/server connections over at [my fork's ruby branch](https://github.com/microlith57/Parade/tree/ruby) and I have thought of some problems of remote warps - `warp to the library@172.20.10.2`, for example.

1. What happens if someone warps themselves to a host, but the host crashes? What would happen to their vessel?
2. What happens if two people possess the same vessel, and *one of them* warps across hosts? What happens to the other client?
3. Regarding `warp the book into the bookshelf`-type warping - should this be allowed across hosts? (eg. `warp the book into the library@172.20.10.2`)
4. What happens if someone warps to a remote host, becomes vessel `#5`, creates a new vessel `#6`, and then disconnects?
   - Should their vessel be deleted, and all subsequent vessels shifted down by 1? This could be expensive on the processor
   - Should their vessel be left behind, causing clutter?
5. What happens if a client warps to a remote host, and the client crashes? Should their vessel be left on the server they warped to?
   - If it is destroyed, their work (eg. passive indicators) may be lost
   - If it is kept, others' libraries may be cluttered with many ghosts
      - Should duplicate names be allowed in this case?

**Comment by microlith57** [comment]
@neauoire Any thoughts on the matter?
If this is not resolved by the time I decide to release v0.1, I think I will temporarily replace this behaviour with a `connect` command (which does not warp vessels) as this will be easier to implement. However I do like your way with the `warp to the vessel@host` syntax so this is the version I will have in v1.0.

**Comment by microlith57** [comment]
Example `connect` syntax:
- `connect to parade://local` - local world file
- `connect to https://192.168.1.14` - local network HTTP
- `disconnect` or `connect to parade://disconnect` - disconnect from server

Connections are stored in a stack - disconnecting pops from the stack, connecting pushes
Client exits when stack is empty
Client interacts with topmost connection
Servers can tell the client to push/pop

**Comment by microlith57** [comment]
In my Ruby client, connections are easy to implement (I did the HTTP one in ~1 hour) - you just do something like this:

```ruby
class MyConnection < Connection
  def greet; end # Retrieve vessel ID, optionally insert vessels into world (via warp)

  def disconnect; end # Tell the server you're going

  def query(string); end # Send a query to the server
end
```

**Comment by neauoire** [comment]
Hi,
sorry for the silence. I'm currently travelling at the moment. I made a clone of your fork to play with in the bus ride :) I'll let you know my thoughts when things calm down and I'm at a place where I can use my laptop.

**Comment by microlith57** [comment]
Thanks! I'll keep working on this in the meantime.

**Comment by neauoire** [comment]
Have you ever used Crystal?

**Comment by microlith57** [comment]
No, I haven't but I just got it and read the docs and it looks _really cool_. I might need to check it out at some point.

**Comment by neauoire** [comment]
> What happens if someone warps themselves to a host, but the host crashes? What would happen to their vessel?

I think that maybe the whole server should be instantiated in the local machine, and exchanges would be like commits on github. Like, you pull someone's world(read only), do some changes and suggest a change to the host.

**Comment by neauoire** [comment]
I've spent the last 2 days doing some pretty serious research in regard to the implementation of Parade. I'm looking for a super simple lisp machine type system that will allow to run some lisp, something like [Interim](https://github.com/mntmn/interim), loaded with the `(parade)` symbol by default that interprets a query inside a REPL.

I find that LISP shares a lot of aesthetics with Paradise, and it would be a good fit. I just need to build a simple base from which we can build.

**Comment by microlith57** [comment]
These are very good ideas! I think a good protocol to serve it with is WebSocket, because it would support telling clients about changes.
Storing the world-state in  a text file would be a good idea if we want to implement a Git-like setup, so it can be passed back and forth when needed.

We could even Actions as a form of Lisp - this would allow for a simpler implementation.

I will mockup a commandline GUI for the clients. I think a 2-pane approach would be good.

**Comment by microlith57** [comment]
I have started a Github Organisation [here](https://github.com/paradise-celestial) for my Crystal implementation - feel free to join if you like.
